lua/entities/ent_jack_gmod_ezsolargenerator.lua

```diff
@@ -0,0 +1,247 @@
+    end
+    
+
+   function ENT:Use(activator)
+		local State=self:GetState()
+		local OldOwner=self.Owner
+		JMod.Owner(self,activator)
+		JMod.Colorify(self)
+		if(IsValid(self.Owner))then
+			if(OldOwner~=self.Owner)then -- if owner changed then reset team color
+				JMod.Colorify(self)
+			end
+		end
+		if(State==STATE_BROKEN)then
+			JMod.Hint(activator,"destroyed",self)
+			return
+		elseif(State==STATE_OFF)then
+			self:TurnOn()
+		elseif(State==STATE_ON)then
+            self:ProducePower()
+			self:TurnOff()
+		end
+	end
+    
+    function ENT:SpawnEffect(pos)
+        local effectdata=EffectData()
+        effectdata:SetOrigin(pos)
+        effectdata:SetNormal((VectorRand()+Vector(0,0,1)):GetNormalized())
+        effectdata:SetMagnitude(math.Rand(5,10))
+        effectdata:SetScale(math.Rand(.5,1.5))
+        effectdata:SetRadius(math.Rand(2,4))
+        util.Effect("Sparks", effectdata)
+        self:EmitSound("items/suitchargeok1.wav", 80, 120)
+    end
+    
+    function ENT:ProducePower()
+        local SelfPos,Up,Forward,Right = self:GetPos(),self:GetUp(),self:GetForward(),self:GetRight()
+        local amt = math.min(math.floor(self:GetProgress()), self.MaxPower)
+
+        if amt <= 0 then return end
+        
+        local pos = SelfPos + Forward*15 - Up*25 - Right*2
+        for _, ent in pairs(ents.FindInSphere(pos, 100)) do
+            --print(ent, ent.GetResourceType and ent:GetResourceType())
+            if ((ent:GetClass() == "ent_jack_gmod_ezcrate") and (ent:GetResourceType() == "generic" 
+            or ent:GetResourceType() == "power") and (ent:GetResource() + amt <= ent.MaxResource)) then
+                    
+                if ent:GetResourceType() == "generic" then
+                    ent:ApplySupplyType("power")
+                end
+                    
+                ent:SetResource(math.min(ent:GetResource() + amt, ent.MaxResource))
+                self:SetProgress(self:GetProgress() - amt)
+                self:SpawnEffect(pos)
+                return
+            end
+        end
+        JMod.MachineSpawnResource(self, "power", amt, self:WorldToLocal(pos), Angle(-90, 0, 0), Up*-300)
+        self:SetProgress(self:GetProgress() - amt)
+        self:SpawnEffect(pos)
+    end
+
+    function ENT:CheckSky()
+        local HitAmount = 0
+        for i = 1, 10 do
+            for j = 1, 10 do
+                local StartPos = self:LocalToWorld(Vector(-5 + j*1, -100 + i*25, 10 + j*7.5))
+                local Dir = self:LocalToWorldAngles(Angle(260 - j*8, -10 + i*2, 0)):Forward()
+                local HitSky = util.TraceLine({start = StartPos, endpos = StartPos + Dir * 9e9, filter = {self}, mask = MASK_SOLID}).HitSky
+                if (HitSky) then HitAmount = HitAmount + 1 end
+                --JMod.Sploom(game.GetWorld(), StartPos + Dir * 1000, 0.5)
+            end
+        end
+        --print(HitAmount)
+        return HitAmount
```

Comment:
i think we shouldn't return a number 0 to 100, instead we should return a float frac 0 to 1
that way, it's easier later to just multiply things by it

Reply:
Rodger doger